### DISCUSSION TITLE
sharing my progress - not a bug - working dockerfile

### DISCUSSION THREAD

**Discussion by ecowitness** [coqui-ai/TTS#569] [Show and tell]
I have temporarily suspended my journey. It turns out that the model I need for Swedish isn't ready yet.

I have a dockerfile that I will share below. I wanted to run this in docker for portability and because of the specific software levels of the requirements. I'm sharing this so that my time won't have been wasted. 

My first build environment was on a raspberry pi. The build choked on pytorch, which I could not resolve, so I moved to x86 64bit.

The dockerfile that I share is the result of several iterations. The simple "pip3 install tts" will stumble multiple times. The setup.py program depends upon several listed requirements being previously installed. One or more of the requirements has a specific level of pip3 that must be installed. So this work-in-progress dockerfile, while nasty in appearance and style, will at least get you to a shell prompt where you can issue commands. I got tts to the point of listing the models and stopped. 

Don't try to run apt-get with --no-install-recommends. While that will make a much smaller image, you will spend a lot of time going back to get individual dependencies. I don't know that python3-dbg is required, but in getting individual requirements to work, it was recommended. Though we often stand on the shoulders of giants, this doesn't necessarily get us a better view.

Note that pip3 will warn about being root, but that shouldn't be a problem in this image, so just shrug off those warnings while building.

There are comments in the dockerfile that I hope will detail some of the work that went into this. 

```
# Dockerfile for coqui
# work in progress
# just to get to the shell prompt in the run

# developer says it works on 18.04 ubuntu 
FROM ubuntu:bionic

# get the image updated 
RUN apt-get update \
  && \
  apt-get upgrade -y 
  
# make sure that
# pip is at a higher version (like > 19)
# or installs will fail
#
# and tts won't start without libsndfile1
RUN  apt-get install -y \
     python3-pip \
     python3-dbg \
     libsndfile1 \
  && \
  pip3 install -U pip

# this bunch includes things that failed
# along the way and they were brought
# here to say that they were troublesome
RUN pip3 install --upgrade setuptools \
  && \
  pip3 install cython \
  && \
  pip3 install numpy==1.18.5 \
  && \
  pip3 install llvmlite==0.35.0 \
  && \
  pip3 install soundfile

# this is basically the requirements file
# from the setup program for tts
# and we brought them here to make sure
# they were installed before we tried tts
# (cython and numpy moved to previous run) 
RUN  pip3 install \
    flask \
    gdown \
    inflect \
    jieba \
    librosa==0.8.0 \
    matplotlib \
    pandas \
    pypinyin \
    pysbd \
    pyyaml \
    scipy>=0.19.0 \
    soundfile \
    tensorboardX \
    torch>=1.7 \
    tqdm \
    numba==0.52 \
    umap-learn==0.4.6 \
    anyascii \
    coqpit \
    mecab-python3==1.0.3 \
    unidic-lite==1.0.8

# the tts setup.py program imports
# numpy and numpy depends upon cython
# so we reorganized the install to make
# sure that stuff was installed before
# we ran the setup.py program
RUN  pip3 install tts \
  && \
  rm -rf /var/lib/apt/lists/* \
  && \
  apt-get -y autoremove 

# there isn't anything in the
# workdir path yet
WORKDIR /

# just the shell, so we can play around in the container
ENTRYPOINT [ "/bin/sh" ]

# build
# sudo docker build -t coqui .
# run
# sudo docker run -it coqui
```

**Comment by rishubil** [comment]
I also share the docker file that I used a few weeks ago.

```dockerfile
FROM pytorch/pytorch:1.7.1-cuda11.0-cudnn8-devel

RUN apt-get update && apt-get install -y \
    build-essential \
    tmux \
    curl \
    unzip \
    git \
    espeak-ng \
    libsndfile1-dev \
 && rm -rf /var/lib/apt/lists/*

RUN mkdir /content
ADD . /content/tts
WORKDIR /content/tts

RUN pip install --no-cache-dir -e .[all]
RUN pip install --no-cache-dir jupyterlab tensorboard

# tts-server
EXPOSE 5002
# tensorboard
EXPOSE 6006
# jupyter lab
EXPOSE 8888


CMD ["jupyter", "lab", "--ip=*", "--port=8888", "--no-browser", "--allow-root"]
```

**Reply by ecowitness** [reply]
Thank you for sharing.

I am new to this project and did not know that there was a docker for pytorch. Thanks. That could have saved me a lot of time and I'll make note of that for next time.

One small point, having to do with portability -- doesn't the "add . /content/tts" imply that tts already exists on the machine where you are doing the docker build?

**Reply by rishubil** [reply]
I added the entire project files to the docker image because I wanted an environment where I could use TTS with just the built docker image.
I built a docker image on my local machine and I used it in VPS with high-performance GPUs.
I didn't do any optimization after checking that it works properly, so please refer to it when you check it.

**Reply by skol101** [reply]
@rishubil 

Cheers for sharing your docker build. A quick question: why not use nvcr.io/nvidia/pytorch:21.08-py3 image?

**Reply by rishubil** [reply]
> @rishubil
> 
> Cheers for sharing your docker build. A quick question: why not use nvcr.io/nvidia/pytorch:21.08-py3 image?

I didn't know there was an image provided by nvidia. It would be nice to use.

**Reply by skol101** [reply]
@rishubil mind sharing which VPS with GPUs have you used? I was trying to rent one from Exoscale.com, but their GPU2 (v100) product is fully booked.

**Reply by rishubil** [reply]
@skol101 I used puzl.ee and iwinv.kr It would be difficult to use iwinv.kr unless you are Korean, so you'd better check puzl.ee
